### PR TITLE
Expose MapReduceTaskContext to MapReduce partitioner/comparator classes

### PR DESCRIPTION
**Pull request by anwar6953**
Basically, allow MapReduce Partitioner/Comparator classes to have access to MapReduceTaskContext. This is done by having Wrapper classes that delegate to the user's implementation of these classes.

See JIRA for more details on use case.

https://issues.cask.co/browse/CDAP-5740
http://builds.cask.co/browse/CDAP-RUT114-3

### CODE REVIEW

**Comment by anwar6953** [comment]
Tested against cluster.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java [hunk 1]
```diff
@@ -111,13 +113,24 @@ protected void shutDown() throws Exception {
    * Returns the {@link BasicMapReduceTaskContext} for the given task.
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java:113]: unrelated to this PR, but I just re-read this code... would it make sense to invalidate the cache on shutdown? To make sure all contexts can be gc'ed even if this provider is not gc'ed immediately?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/PartitionerWrapper.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+package co.cask.cdap.internal.app.runtime.batch;
+
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.MRJobConfig;
+import org.apache.hadoop.mapreduce.Partitioner;
+
+/**
+ * Wraps user-defined implementation of {@link Partitioner} class which allows perform extra configuration.
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/PartitionerWrapper.java]: allows to perform

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java [hunk 3]
```diff
@@ -207,10 +229,15 @@ public BasicMapReduceTaskContext load(ContextCacheKey key) throws Exception {
     private final Configuration configuration;
 
     private ContextCacheKey(TaskAttemptContext context) {
-      this.taskAttemptID = context.getTaskAttemptID();
-      this.configuration = context.getConfiguration();
+      this(context.getTaskAttemptID(), context.getConfiguration());
+    }
+
+    private ContextCacheKey(@Nullable TaskAttemptID taskAttemptID, Configuration configuration) {
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java:235]: Am I understanding this correctly? 

For a partitioner or a comparator, the taskAttemptId will be null. That is, there is no metrics collection, right? 

Also, in standalone, the partitioner and comparator are instantiated by every mapper, is that right? If so, then that instance is cached and all mappers will share the same instance of the context. Can that lead to problems? For example, would it get initialized multiple times?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/RawComparatorWrapper.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+package co.cask.cdap.internal.app.runtime.batch;
+
+import org.apache.hadoop.conf.Configurable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.io.RawComparator;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.MRJobConfig;
+
+/**
+ * Wraps user-defined implementation of {@link RawComparator} class which allows perform extra configuration,
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/RawComparatorWrapper.java]: allows to perform

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/RawComparatorWrapper.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    return delegate.compare(b1, s1, l1, b2, s2, l2);
+  }
+
+  @Override
+  public int compare(Object o1, Object o2) {
+    return delegate.compare(o1, o2);
+  }
+
+  static final class CombinerGroupComparatorWrapper extends RawComparatorWrapper {
+    private static final String ATTR_PARTITIONER_CLASS = "c.combiner.group.comparator.class";
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/RawComparatorWrapper.java]: misleading constant name? same for the next two classes.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/WrapperUtil.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.util.ReflectionUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/WrapperUtil.java]: empty Javadoc?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/WrapperUtil.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      return delegate;
+    }
+
+    MapReduceClassLoader classLoader = MapReduceClassLoader.getFromConfiguration(conf);
+    BasicMapReduceTaskContext basicMapReduceContext = classLoader.getTaskContextProvider().get(conf);
+    ClassLoader programClassLoader = classLoader.getProgramClassLoader();
+
+    ClassLoader oldClassLoader = ClassLoaders.setContextClassLoader(programClassLoader);
+    try {
+      ((ProgramLifecycle) delegate).initialize(new MapReduceLifecycleContext(basicMapReduceContext));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/WrapperUtil.java]: Is it true that we only call initialize() but never destroy()?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/PartitionerWrapper.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    return delegate.getPartition(o, o2, numPartitions);
+  }
+
+  @Override
+  public void setConf(Configuration conf) {
+    this.conf = conf;
+    this.delegate = WrapperUtil.createDelegate(conf, ATTR_PARTITIONER_CLASS);
+  }
+
+  @Override
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/PartitionerWrapper.java:56]: If the delegate is Configurable(), then we should delegate this to it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/RawComparatorWrapper.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+  @Override
+  public void setConf(Configuration conf) {
+    this.conf = conf;
+    this.delegate = WrapperUtil.createDelegate(conf, getDelegateClassAttr());
+  }
+
+  @Override
+  public Configuration getConf() {
+    return conf;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/RawComparatorWrapper.java:44]: Same as for the partitioner wrapper. If the delegate is Configurable(), then we should delegate this to it.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/WrapperUtil.java (hunk 7 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/WrapperUtil.java]: if the delegate is a Configurable, then we need to call its setConf() method, too.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java [hunk 10]
```diff
@@ -62,7 +61,6 @@
   protected final TransactionSystemClient txClient;
   protected final NamespaceId namespace;
   protected final Map<String, String> runtimeArguments;
-  protected final MetricsContext metricsContext;
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java:65]: is this an unrelated change? Is it because this is never used by the class?

**Comment by anew** [comment]
Does this need documentation? Especially, that the destroy() is never called.

**Comment by anew** [comment]
Is there really now way to destroy? Perhaps by remembering it in the MRTaskContext, so when that is closed, it can call destroy()? Or would that be super complicated?

**Comment by anew** [comment]
@anwar6953  I made a pass through this, please see my comments.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java (hunk 10 above)
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/DynamicDatasetCache.java:65]: yea; its never used

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java (hunk 3 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java:235]: Right - for partitioner or comparator, there will be no metrics collections.

In Standalone then, the Partitioners and Comparators would share a MapReduceTaskContext. I don't see how that would be an issue.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/RawComparatorWrapper.java (hunk 5 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/RawComparatorWrapper.java]: fixed (made more generic)

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/WrapperUtil.java (hunk 7 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/WrapperUtil.java]: Yea, we don't call `destroy`.

`ReflectionsUtils#newInstance` takes care of calling the delegate's `setConf` method.

**Comment by anwar6953** [comment]
Yea, there will need to be documentation for this.

**Comment by anwar6953** [comment]
Good idea on calling its `destroy` methods. Updated and pushed code to do so.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java (hunk 3 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java:235]: Regarding metrics collection, let's make sure it's documented. What would happen if the user attempts to emit metrics anyway?

It might not be a problem that the context would be shared. I was just wondering whether there can be anything troublesome. For example, if they getDataset(), they would share that instance of the dataset, right?

Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java:235]: I'm not sure metrics is available to them there. I'll check.

**Comment by anew** [comment]
New changes look good. There are still a couple of questions: 
- getConf() if the delegate is Configurable?
- sharing of MapReduceTaskContext
- docs

**Comment by anwar6953** [comment]
I checked that Metrics are not exposed via MapReduceTaskContext.

I also added test cases to show that `setConf` is being called if the delegate is `Configurable`. This is handled by `ReflectionUtils.newInstance`.

Filed a JIRA for documentation: https://issues.cask.co/browse/CDAP-7109 (can be done next week).

I verified that the MapReduceTaskContext is shared across the Comparators and Partitioners.
They would share the same instance of the dataset, but is there any issue with that?

**Comment by anew** [comment]
Sharing a dataset is an issue because that would be across threads, and datasets are not thread safe. I am not sure whether that is a typical use for datasets, but since you can't emit metrics for the partitioner/comparator, if you want to record somehow - say - the distribution of the partitioning to get an idea about skew, you would have to do it by counting in memory and writing to a dataset from the destroy method. Now if multiple threads write using the same dataset, they'll probably cancel out each other's writes. Perhaps there could even be bigger problems like concurrent modification exceptions, I am not sure, but I know that the Table implementation is not thread-safe.   

What was the use case for exposing the MapReduceTaskContext to these classes?

**Comment by anew** [comment]
+1 for the latest changes. Let's discuss the sharing of contexts between instances to get this merged.

**Comment by anwar6953** [comment]
Use case is being able to access workflow token and datasets for partitioning logic.

**Comment by anwar6953** [comment]
Pushed a change to make BasicMapReduceTaskContext use MultiThreadedDatasetCache.

**Comment by anwar6953** [comment]
Filed a JIRA to document limitations of datasets:
https://issues.cask.co/browse/CDAP-7111

**Comment by anew** [comment]
LGTM after build passes.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java (hunk 1 above)
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceTaskContextProvider.java:113]: I don't think gc is a concern since on shutdown, the JVM is terminating anyways.

**Comment by anwar6953** [comment]
Build is green. Squashed and merging.
